src/psyclone/psyGen.py

```diff
@@ -5356,13 +5538,112 @@ def _if_stmt_handler(self, node, parent):
+            if isinstance(child, (Fortran2003.Case_Stmt,
+                                  Fortran2003.End_Select_Stmt)):
+                clause_indices.append(idx)
+                # Case Default and value Ranges not supported yet
+                if 'DEFAULT' in str(child):
+                    raise NotImplementedError("Case Default Statement")
+                if ':' in str(child):
+                    raise NotImplementedError("Case Value Range Statement")
+
+        # Deal with each Case_Stmt
+        rootif = None
+        currentparent = parent
+        num_clauses = len(clause_indices) - 1
+        for idx in range(num_clauses):
+            start_idx = clause_indices[idx]
+            end_idx = clause_indices[idx+1]
+            clause = node.content[start_idx]
+
+            if isinstance(clause, Fortran2003.Case_Stmt):
+                case = clause.items[0]
+                if isinstance(case, Fortran2003.Case_Selector):
+                    ifblock = IfBlock(parent=currentparent,
+                                      annotation='was_case')
+
+                    # Add condition: selector == case
+                    bop = BinaryOperation(parent=ifblock, operator='==')
+                    self.process_nodes(parent=bop,
+                                       nodes=[selector],
+                                       nodes_parent=node)
+                    self.process_nodes(parent=bop,
+                                       nodes=[case.items[0]],
+                                       nodes_parent=node)
+                    ifblock.addchild(bop)
+
+                    # Add If_body
+                    ifbody = Schedule(parent=ifblock)
+                    self.process_nodes(parent=ifbody,
+                                       nodes=node.content[start_idx + 1:
+                                                          end_idx],
+                                       nodes_parent=node)
+                    ifblock.addchild(ifbody)
+                    # Keep pointer to fpaser2 AST
+                    ifbody._ast = node.content[start_idx]
+                    ifbody._ast_end = node.content[end_idx]
+
+                    if rootif:
+                        # If rootif is already initialised we chain the new
+                        # case in the last else branch.
+                        elsebody = Schedule(parent=currentparent)
+                        currentparent.addchild(elsebody)
+                        elsebody.addchild(ifblock)
+                        # Keep pointer to fpaser2 AST
+                        elsebody._ast = node.content[start_idx]
+                        ifblock._ast = node.content[start_idx]
+                        elsebody._ast_end = node.content[end_idx]
+                        ifblock._ast_end = node.content[end_idx]
+                    else:
+                        rootif = ifblock
+                        rootif._ast = node  # Keep pointer to fpaser2 AST
```

Comment:
Is this correct? It doesn't seem to agree with how you set `ifblock._ast` just above. Also, please could you set `rootif._ast_end`. Can this setting of `ifblock._ast/_ast_end` be done outside the `if rootif` block?

Reply:
I updated this code appropriately, now `ifblock._ast/_ast_end` is common

Reply:
Great, thanks.